core/base/v7/inc/ROOT/TDirectoryEntry.hxx

```diff
@@ -33,7 +31,7 @@ public:
 
 private:
   time_point_t fDate = clock_t::now(); ///< Time of last change
-  TClass* fType;
+  const std::type_info* fType;
```

Comment:
TClass::DynamicCast is likely to be (much?) slower than the code you very nicely introduced.  Removing/Reducing the usage of TClass is an end goal that requires to remove each part/usage one by one :)